research/eccv2024/export_to_webdataset_and_crops.py

```diff
@@ -0,0 +1,215 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+"""
+Author        : Aditya Jain
+Date started  : April 9, 2024
+About         : Exports the AMI-Traps dataset to webdataset format and individual crops
+"""
+
+import os
+import json
+from PIL import Image
+import webdataset as wds
+import pandas as pd
+
+
+def get_synonym(sp_checklist: pd.DataFrame, species: str):
+    """Return synonym name of a species, if it exists on GBIF"""
+
+    # Search for species in the checklist
+    try:
+        synonym = sp_checklist.loc[sp_checklist["search_species"] == species][
+            "gbif_species"
+        ].values[0]
+    except:
+        synonym = None
+        # print(f"{species} is not found in the search species column.", flush=True)
+
+    if synonym == species:
+        return None
+    else:
+        return synonym
+
+
+def get_gbif_keys(sp_key_map: pd.DataFrame, sp_checklist: pd.DataFrame, species: str):
+    """Get the species key and accepted taxon key for a given species"""
+
+    speciesKey, acceptedTaxonKey = None, None
+
+    # Get species key from GBIF species key map
+    try:
+        speciesKey = int(
+            sp_key_map.loc[sp_key_map["species"] == species]["speciesKey"].values[0]
+        )
+    except:
+        pass
+
+    # Get accepted taxon key from checklist
+    try:
+        acceptedTaxonKey = int(
+            sp_checklist.loc[sp_checklist["search_species"] == species][
+                "accepted_taxon_key"
+            ].values[0]
+        )
+    except:
+        try:
+            acceptedTaxonKey = int(
+                sp_checklist.loc[sp_checklist["gbif_species"] == species][
+                    "accepted_taxon_key"
+                ].values[0]
+            )
+        except:
+            pass
+
+    return speciesKey, acceptedTaxonKey
+
+
+def export_to_webdataset_and_crops(
+    data_dir: str, export_dir: str, sp_checklist: pd.DataFrame, sp_key_map: pd.DataFrame
+):
+    """Main function for exporting AMI-Traps to webdataset and individual crops"""
+
+    # Get the list of raw camera trap images and other metadata
+    image_dir = os.path.join(data_dir, "images")
+    metadata_dir = os.path.join(data_dir, "metadata")
+    labels_dir = os.path.join(data_dir, "labels")
+    image_list = os.listdir(image_dir)
+
+    # Get ground truth class names
+    gt_class_list = json.load(open(os.path.join(data_dir, "notes.json")))["categories"]
+
+    # Webdataset specific variables
+    max_shard_size = 50 * 1024 * 1024
+    wbd_pattern_binary = os.path.join(
+        export_dir, "webdataset", "binary_classification", "binary-%06d.tar"
+    )
+    wbd_pattern_fgrained = os.path.join(
+        export_dir, "webdataset", "fine-grained_classification", "fgrained-%06d.tar"
+    )
+    sink_binary = wds.ShardWriter(wbd_pattern_binary, maxsize=max_shard_size)
+    sink_fgrained = wds.ShardWriter(wbd_pattern_fgrained, maxsize=max_shard_size)
```

Comment:
I'm pretty sure you could use a `with` statement, as was done here instead of manually opening and closing : https://github.com/RolnickLab/ami-ml/blob/main/src/dataset_tools/create_webdataset.py#L290

If you are unfamilliar with this, I suggest this resource :https://realpython.com/python-with-statement/#the-with-statement-approach

Reply:
@f-PLT I could but I need to write the same image in two sinks simultaneously which I am unsure if could be done using a `with` statement. Am I missing something?

Reply:
You can use a `with` in a `with`

```python
with something as bob:
    with something_else as mike:
        <do something with bob>
        <do something with mike>
```

Reply:
Done in d56f425.